source/fundamentals/logging.txt

```diff
@@ -0,0 +1,273 @@
+      driver uses the Logback framework to output your logs.
+
+      .. code-block:: xml
+
+         <!-- SLF4J API -->
+         <dependency>
+           <groupId>org.slf4j</groupId>
+           <artifactId>slf4j-api</artifactId>
+           <version>1.7.30</version>
+         </dependency>
+         <!-- This package contains an SLF4J-Logback binding -->
+         <dependency>
+           <groupId>ch.qos.logback</groupId>
+           <artifactId>logback-classic</artifactId>
+           <version>1.2.3</version>
+         </dependency>
+         <!-- Logback Core Functionality -->
+         <dependency>
+           <groupId>ch.qos.logback</groupId>
+           <artifactId>logback-core</artifactId>
+           <version>1.2.3</version>
+         </dependency>
+
+      Once you have installed the above dependencies, try connecting to a
+      MongoDB instance and retrieving a document with the following code.
+
+      .. code-block:: java
+
+         MongoClient mongoClient = MongoClients.create(<my uri>);
+         MongoDatabase database = mongoClient.getDatabase(<my database>);
+         MongoCollection<Document> collection = database.getCollection(<my collection>);
+         collection.find().first();
+
+      You should see output resembling the following:
+
+      .. code-block:: none
+
+         ...
+         1156 [main] DEBUG o.m.driver.protocol.command - Sending command '{"find": "<your collection>", "filter": {}, "limit": 1, "singleBatch": true, "$db": "<your db>", "lsid": {"id": {"$binary": {"base64": "<server session>", "subType": "04"}}}}' with request id 13 to database <your db> on connection <your connection> to server <your server>
+         1173 [main] DEBUG o.m.driver.protocol.command - Execution of command with request id 13 completed successfully in 30.01 ms on connection <your connection> to server <your server>
+   
+      For more information on Logback, see the
+      `Logback manual <http://logback.qos.ch/manual/>`__.
+   
+   .. tab:: Log4j2
+      :tabid: Log4j2-binding
+   
+      If you add these dependencies to your ``pom.xml`` file, the MongoDB Java
+      driver uses the Log4j2 framework to output your logs.
+
+      .. code-block:: xml      
+      
+         <!-- SLF4J API -->
+         <dependency>
+           <groupId>org.slf4j</groupId>
+           <artifactId>slf4j-api</artifactId>
+           <version>1.7.30</version>
+         </dependency>
+         <!-- SLF4J-Log4j2 binding -->
+         <dependency>
+           <groupId>org.apache.logging.log4j</groupId>
+           <artifactId>log4j-slf4j-impl</artifactId>
+           <version>2.14.1</version>
+         </dependency>
+         <!-- Log4j2 core functionality -->
+         <dependency>
+           <groupId>org.apache.logging.log4j</groupId>
+           <artifactId>log4j-core</artifactId>
+           <version>2.14.1</version>
+         </dependency>
+
+      Once you have installed the above dependencies, try logging an error.
```

Comment:
Why is this step different with this framework?

Reply:
Log4J2 has a default logger level of "[ERROR](https://logging.apache.org/log4j/2.x/manual/configuration.html)", where as Logback has a default logger level of "[DEBUG](https://github.com/qos-ch/logback/blob/master/logback-examples/src/main/resources/chapters/configuration/sample0.xml)" . This means that performing basic operations with the driver will give output for Logback but not Log4j2. There is a way to change the root logger level of Log4J2 [programmatically](https://logging.apache.org/log4j/log4j-2.4/faq.html#reconfig_level_from_code), but it is not supported by the public api. Configuring the log framework to change the log level through a config file is not easy, and is the subject of the next section, so I thought it made sense to use different examples (you can only use 1 log framework at a time as well). 

I tried finding a way to make the driver output an "ERROR" log without using esoteric code, but had trouble doing it (Maybe Jeff Yemin would know a way?).